### ISSUE TITLE
Multiple problems

### ISSUE THREAD

**Issue by DuncanBetts** [miguelgrinberg/Flask-SocketIO#487]
Hi

Here are some of the error messages I see:

```
website_1   | 2017-06-25 11:43:31,367 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': '07ef445f29e34b5bb94197b13f5adf55'}
website_1   | 2017-06-25 11:43:31,368 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Sending packet MESSAGE data 0
website_1   | 2017-06-25 11:43:31,458 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Received packet MESSAGE data 0/notifs
website_1   | 2017-06-25 11:43:31,458 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Sending packet MESSAGE data 0/notifs
website_1   | 2017-06-25 11:43:31,472 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Received request to upgrade to websocket
website_1   | 2017-06-25 11:43:31,475 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Sending packet NOOP data None
website_1   | 2017-06-25 11:43:31,481 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Received packet MESSAGE data 2/notifs,["message","User has arrived on the progress page."]
website_1   | 2017-06-25 11:43:31,483 10 INFO Anon socketio received event "message" from 07ef445f29e34b5bb94197b13f5adf55 [/notifs]
website_1   | 2017-06-25 11:43:31,497 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Upgrade to websocket successful
website_1   | 2017-06-25 11:43:31,499 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Received packet MESSAGE data 2/notifs,["join_part_room",{"part_id":"1105"}]
website_1   | 2017-06-25 11:43:31,500 10 INFO Anon socketio received event "join_part_room" from 07ef445f29e34b5bb94197b13f5adf55 [/notifs]
website_1   | 2017-06-25 11:43:32,145 10 INFO Anon engineio d2d290e74cc04c3f85715a0037d0d538: Sending packet OPEN data {'pingInterval': 25000, 'pingTimeout': 60000, 'upgrades': ['websocket'], 'sid': 'd2d290e74cc04c3f85715a0037d0d538'}
website_1   | 2017-06-25 11:43:32,145 10 INFO Anon engineio d2d290e74cc04c3f85715a0037d0d538: Sending packet MESSAGE data 0
website_1   | 2017-06-25 11:43:32,542 10 INFO Anon engineio d2d290e74cc04c3f85715a0037d0d538: Received packet MESSAGE data 0/notifs
website_1   | 2017-06-25 11:43:32,542 10 INFO Anon engineio d2d290e74cc04c3f85715a0037d0d538: Sending packet MESSAGE data 0/notifs
website_1   | 2017-06-25 11:43:32,856 10 INFO Anon engineio d2d290e74cc04c3f85715a0037d0d538: Received request to upgrade to websocket
website_1   | 2017-06-25 11:43:33,195 10 INFO Anon engineio d2d290e74cc04c3f85715a0037d0d538: Sending packet NOOP data None
website_1   | 2017-06-25 11:43:33,507 10 INFO Anon engineio d2d290e74cc04c3f85715a0037d0d538: Failed websocket upgrade, expected UPGRADE packet, received None instead.
```

```
website_1   | 2017-06-25 11:44:33,347 10 INFO Anon engineio d2d290e74cc04c3f85715a0037d0d538: Client is gone, closing socket
website_1   | [2017-06-25 11:44:33 +0000] [10] [ERROR] Error handling request /socket.io/?EIO=3&transport=polling&t=1498391013256-4&sid=d2d290e74cc04c3f85715a0037d0d538
website_1   | Traceback (most recent call last):
website_1   |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 56, in handle
website_1   |     self.handle_request(listener_name, req, client, addr)
website_1   |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 107, in handle_request
website_1   |     respiter = self.wsgi(environ, resp.start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1994, in __call__
website_1   |     return self.wsgi_app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/sqltap/wsgi.py", line 42, in __call__
website_1   |     return self.app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/werkzeug/contrib/fixers.py", line 152, in __call__
website_1   |     return self.app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/flask_socketio/__init__.py", line 42, in __call__
website_1   |     start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/middleware.py", line 47, in __call__
website_1   |     return self.engineio_app.handle_request(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/socketio/server.py", line 353, in handle_request
website_1   |     return self.eio.handle_request(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/server.py", line 268, in handle_request
website_1   |     self.disconnect(sid)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/server.py", line 202, in disconnect
website_1   |     self._get_socket(sid).close()
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/server.py", line 411, in _get_socket
website_1   |     raise KeyError('Session is disconnected')
website_1   | KeyError: 'Session is disconnected'
```

```
website_1   | 2017-06-25 11:44:36,207 10 INFO Anon engineio 07ef445f29e34b5bb94197b13f5adf55: Unexpected error "[Errno 107] Transport endpoint is not connected", closing connection
```

I have enabled logging, and earlier today added the message_queue parameter too, which hasn't helped (doesn't seem to have changed anything):

```
socketio.init_app(app, message_queue='redis://:devpassword@redis:6379/1',
                             logger=True, engineio_logger=True)
```

pip package versions (excerpt):
```
eventlet==0.21.0 (have also tried 0.20.1, 0.18)
Flask==0.12
Flask-Socketio==2.8.6
gevent==1.2.1 (tried removing this to see if it helped, it did not)
greenlet==0.4.12
gunicorn==19.7.1
python-engineio==1.5.4
python-socketio==1.7.5
redis==2.10.5
```

Here's another error I see:

```
website_1   | [2017-06-25 12:36:12 +0000] [10] [ERROR] Socket error processing request.
website_1   | Traceback (most recent call last):
website_1   |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 66, in handle
website_1   |     six.reraise(*sys.exc_info())
website_1   |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 56, in handle
website_1   |     self.handle_request(listener_name, req, client, addr)
website_1   |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 129, in handle_request
website_1   |     six.reraise(*sys.exc_info())
website_1   |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 107, in handle_request
website_1   |     respiter = self.wsgi(environ, resp.start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1994, in __call__
website_1   |     return self.wsgi_app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/sqltap/wsgi.py", line 42, in __call__
website_1   |     return self.app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/werkzeug/contrib/fixers.py", line 152, in __call__
website_1   |     return self.app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/flask_socketio/__init__.py", line 42, in __call__
website_1   |     start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/middleware.py", line 47, in __call__
website_1   |     return self.engineio_app.handle_request(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/socketio/server.py", line 353, in handle_request
website_1   |     return self.eio.handle_request(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/server.py", line 261, in handle_request
website_1   |     environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/socket.py", line 89, in handle_get_request
website_1   |     start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/socket.py", line 136, in _upgrade_websocket
website_1   |     return ws(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/async_eventlet.py", line 19, in __call__
website_1   |     return super(WebSocketWSGI, self).__call__(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/websocket.py", line 127, in __call__
website_1   |     self.handler(ws)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/socket.py", line 161, in _websocket_handler
website_1   |     pkt = ws.wait()
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/websocket.py", line 633, in wait
website_1   |     for i in self.iterator:
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/websocket.py", line 503, in _iter_frames
website_1   |     message = self._recv_frame(message=fragmented_message)
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/websocket.py", line 526, in _recv_frame
website_1   |     header = recv(2)
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/websocket.py", line 442, in _get_bytes
website_1   |     d = self.socket.recv(numbytes - len(data))
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/greenio/base.py", line 360, in recv
website_1   |     return self._recv_loop(self.fd.recv, b'', bufsize, flags)
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/greenio/base.py", line 354, in _recv_loop
website_1   |     self._read_trampoline()
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/greenio/base.py", line 325, in _read_trampoline
website_1   |     timeout_exc=socket_timeout('timed out'))
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/greenio/base.py", line 207, in _trampoline
website_1   |     mark_as_closed=self._mark_as_closed)
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/__init__.py", line 163, in trampoline
website_1   |     return hub.switch()
website_1   |   File "/usr/local/lib/python2.7/site-packages/eventlet/hubs/hub.py", line 295, in switch
website_1   |     return self.greenlet.switch()
website_1   | timeout: timed out
```

Importantly, my event handlers sometimes still run, if no-one is listening in a room, for example:

```
website_1   | 2017-06-25 12:37:11,958 10 INFO Anon sqlalchemy.engine.base.Engine {'param_1': 1107}
website_1   | --------------------------------------------------------------------------------
website_1   | INFO in events [/flask_app/flask_app/blueprints/quote/events.py:60]:
website_1   | part_saved_notification_ran
website_1   | --------------------------------------------------------------------------------
website_1   | 2017-06-25 12:37:11,963 10 INFO Anon socketio emitting event "part_saved" to admin [/notifs]
```

I'm instantiating gunicorn with:

`gunicorn --bind:'0.0.0.0:8000' --workers=1 --worker-class='eventlet' --log-file='-' "flask_app.app:create_app()"`

**Comment by miguelgrinberg** [comment]
Can I ask you to install `python-engineio==1.6.0` and retry? There was an issue in v1.5.4 that caused errors in your own application to not be reported, and instead you would get the `session is disconnected` error you show above.

Also note that having both eventlet and gevent installed does not really do much. You have to pick one and uninstall the other. Since you seem to be using eventlet, I would uninstall gevent from your virtualenv.

The message queue option is only used when you have multiple server processes. If you have only one server, then it's just going to make things a bit slower without giving you any benefit.

To summarize, what I'd like to see is if doing the upgrade above helps with your issues.

**Comment by DuncanBetts** [comment]
Hi, thanks for the quick response, really appreciated.

I have followed your recommendations; upgraded python-engineio to 1.6.0, removed gevent, and removed the message queue parameter.

I think I'm getting more information about the "session is disconnected" error.

```
website_1   | 2017-06-26 11:02:03,700 11 INFO Anon engineio 4bbe94d092924a8684814a29a1ef156f: Upgrade to websocket successful
website_1   | 2017-06-26 11:02:03,866 11 INFO Anon engineio 41cedc851a55446da10cf1be767b812f: Client is gone, closing socket
website_1   | [2017-06-26 11:02:03 +0000] [11] [ERROR] Error handling request /socket.io/?EIO=3&transport=polling&t=1498474863863-6&sid=41cedc851a55446da10cf1be767b812f
website_1   | Traceback (most recent call last):
website_1   |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 56, in handle
website_1   |     self.handle_request(listener_name, req, client, addr)
website_1   |   File "/usr/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 107, in handle_request
website_1   |     respiter = self.wsgi(environ, resp.start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/flask/app.py", line 1994, in __call__
website_1   |     return self.wsgi_app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/sqltap/wsgi.py", line 42, in __call__
website_1   |     return self.app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/werkzeug/contrib/fixers.py", line 152, in __call__
website_1   |     return self.app(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/flask_socketio/__init__.py", line 42, in __call__
website_1   |     start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/middleware.py", line 47, in __call__
website_1   |     return self.engineio_app.handle_request(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/socketio/server.py", line 353, in handle_request
website_1   |     return self.eio.handle_request(environ, start_response)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/server.py", line 268, in handle_request
website_1   |     self.disconnect(sid)
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/server.py", line 202, in disconnect
website_1   |     self._get_socket(sid).close()
website_1   |   File "/usr/local/lib/python2.7/site-packages/engineio/server.py", line 408, in _get_socket
website_1   |     raise KeyError('Session is disconnected')
website_1   | KeyError: 'Session is disconnected'
```

I've been trying to think if there's information I could instruct my app to output to the logger that might be helpful.

**Comment by DuncanBetts** [comment]
Hi

That occurred because I closed the browser tab that client was using.

I've been asked to leave the functionality that was being powered by Flask-SocketIO as is (broken; it worked previously) and for now switch my focus to other parts of the project.

Thanks for your help so far.

**Comment by DuncanBetts** [comment]
Closing this for now. Still having problems, but I have the impression it's my almost certainly implementation that is at fault, rather than Flask-SocketIO, and this is enough for me to go on - I'll create a barebones flask app and try and figure out how this is happening.
